### ISSUE TITLE
LoweringError after NotImplementedError

### ISSUE THREAD

**Issue by cstich** [numba/numba#3848]
I don't really know what's happening so I am hard pressed to describe it in detail.  I have a minimal working example though.

I am using:
```
python --version
Python 3.7.2

pipenv --version
pipenv, version 11.10.1

pip freeze | ag numba
numba==0.42.0

pip freeze | ag numpy
numpy==1.15.4

pip freeze | ag jupyter
jupyter==1.0.0
jupyter-client==5.2.4
jupyter-console==6.0.0
jupyter-core==4.4.0
```

Minimal working example:
````
import numpy as np
from numba import njit

@njit
def truth_value():
    bar = np.zeros(4)
    if bar:
        bar[1] = 1
        return bar
    else:
        return bar
    
truth_value()
````

Traceback:
````
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/targets/base.py in cast(self, builder, val, fromty, toty)
    689             impl = self._casts.find((fromty, toty))
--> 690             return impl(self, builder, fromty, toty, val)
    691         except NotImplementedError:

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/targets/numbers.py in any_to_boolean(context, builder, fromty, toty, val)
   1385 def any_to_boolean(context, builder, fromty, toty, val):
-> 1386     return context.is_true(builder, fromty, val)
   1387 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/targets/base.py in is_true(self, builder, typ, val)
    725         """
--> 726         impl = self.get_function(bool, typing.signature(types.boolean, typ))
    727         return impl(builder, (val,))

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/targets/base.py in get_function(self, fn, sig, _firstcall)
    550             self.refresh()
--> 551             return self.get_function(fn, sig, _firstcall=False)
    552 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/targets/base.py in get_function(self, fn, sig, _firstcall)
    552 
--> 553         raise NotImplementedError("No definition for lowering %s%s" % (key, sig))
    554 

NotImplementedError: No definition for lowering <class 'bool'>(array(float64, 1d, C),) -> bool

During handling of the above exception, another exception occurred:

NotImplementedError                       Traceback (most recent call last)
~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    616     try:
--> 617         yield
    618     except NumbaError as e:

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/lowering.py in lower_block(self, block)
    258                                    loc=self.loc, errcls_=defaulterrcls):
--> 259                 self.lower_inst(inst)
    260 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/lowering.py in lower_inst(self, inst)
    316             condty = self.typeof(inst.cond.name)
--> 317             pred = self.context.cast(self.builder, cond, condty, types.boolean)
    318             assert pred.type == Type.int(1), ("cond is not i1: %s" % pred.type)

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/targets/base.py in cast(self, builder, val, fromty, toty)
    692             raise NotImplementedError(
--> 693                 "Cannot cast %s to %s: %s" % (fromty, toty, val))
    694 

NotImplementedError: Cannot cast array(float64, 1d, C) to bool: %".73" = load {i8*, i8*, i64, i64, double*, [1 x i64], [1 x i64]}, {i8*, i8*, i64, i64, double*, [1 x i64], [1 x i64]}* %"bar"

During handling of the above exception, another exception occurred:

LoweringError                             Traceback (most recent call last)
<ipython-input-3-127d50b78b35> in <module>
      8         return bar
      9 
---> 10 truth_value()

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    365                     e.patch_message(''.join(e.args) + help_msg)
    366             # ignore the FULL_TRACEBACKS config, this needs reporting!
--> 367             raise e
    368 
    369     def inspect_llvm(self, signature=None):

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    322                 argtypes.append(self.typeof_pyval(a))
    323         try:
--> 324             return self.compile(tuple(argtypes))
    325         except errors.TypingError as e:
    326             # Intercept typing error that may be due to an argument

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler_lock.py in _acquire_compile_lock(*args, **kwargs)
     30         def _acquire_compile_lock(*args, **kwargs):
     31             with self:
---> 32                 return func(*args, **kwargs)
     33         return _acquire_compile_lock
     34 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/dispatcher.py in compile(self, sig)
    653 
    654             self._cache_misses[sig] += 1
--> 655             cres = self._compiler.compile(args, return_type)
    656             self.add_overload(cres)
    657             self._cache.save_overload(sig, cres)

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/dispatcher.py in compile(self, args, return_type)
     80                                       args=args, return_type=return_type,
     81                                       flags=flags, locals=self.locals,
---> 82                                       pipeline_class=self.pipeline_class)
     83         # Check typing error if object mode is used
     84         if cres.typing_error is not None and not flags.enable_pyobject:

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library, pipeline_class)
    924     pipeline = pipeline_class(typingctx, targetctx, library,
    925                               args, return_type, flags, locals)
--> 926     return pipeline.compile_extra(func)
    927 
    928 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in compile_extra(self, func)
    372         self.lifted = ()
    373         self.lifted_from = None
--> 374         return self._compile_bytecode()
    375 
    376     def compile_ir(self, func_ir, lifted=(), lifted_from=None):

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in _compile_bytecode(self)
    855         """
    856         assert self.func_ir is None
--> 857         return self._compile_core()
    858 
    859     def _compile_ir(self):

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in _compile_core(self)
    842         self.define_pipelines(pm)
    843         pm.finalize()
--> 844         res = pm.run(self.status)
    845         if res is not None:
    846             # Early pipeline completion

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler_lock.py in _acquire_compile_lock(*args, **kwargs)
     30         def _acquire_compile_lock(*args, **kwargs):
     31             with self:
---> 32                 return func(*args, **kwargs)
     33         return _acquire_compile_lock
     34 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in run(self, status)
    253                     # No more fallback pipelines?
    254                     if is_final_pipeline:
--> 255                         raise patched_exception
    256                     # Go to next fallback pipeline
    257                     else:

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in run(self, status)
    244                 try:
    245                     event(stage_name)
--> 246                     stage()
    247                 except _EarlyPipelineCompletion as e:
    248                     return e.result

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in stage_nopython_backend(self)
    715         """
    716         lowerfn = self.backend_nopython_mode
--> 717         self._backend(lowerfn, objectmode=False)
    718 
    719     def stage_compile_interp_mode(self):

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in _backend(self, lowerfn, objectmode)
    664             self.library.enable_object_caching()
    665 
--> 666         lowered = lowerfn()
    667         signature = typing.signature(self.return_type, *self.args)
    668         self.cr = compile_result(

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in backend_nopython_mode(self)
    651                 self.calltypes,
    652                 self.flags,
--> 653                 self.metadata)
    654 
    655     def _backend(self, lowerfn, objectmode):

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/compiler.py in native_lowering_stage(targetctx, library, interp, typemap, restype, calltypes, flags, metadata)
   1046         lower = lowering.Lower(targetctx, library, fndesc, interp,
   1047                                metadata=metadata)
-> 1048         lower.lower()
   1049         if not flags.no_cpython_wrapper:
   1050             lower.create_cpython_wrapper(flags.release_gil)

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/lowering.py in lower(self)
    176         if self.generator_info is None:
    177             self.genlower = None
--> 178             self.lower_normal_function(self.fndesc)
    179         else:
    180             self.genlower = self.GeneratorLower(self)

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/lowering.py in lower_normal_function(self, fndesc)
    217         # Init argument values
    218         self.extract_function_arguments()
--> 219         entry_block_tail = self.lower_function_body()
    220 
    221         # Close tail of entry block

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/lowering.py in lower_function_body(self)
    242             bb = self.blkmap[offset]
    243             self.builder.position_at_end(bb)
--> 244             self.lower_block(block)
    245 
    246         self.post_lower()

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/lowering.py in lower_block(self, block)
    257             with new_error_context('lowering "{inst}" at {loc}', inst=inst,
    258                                    loc=self.loc, errcls_=defaulterrcls):
--> 259                 self.lower_inst(inst)
    260 
    261     def create_cpython_wrapper(self, release_gil=False):

~/.pyenv/versions/3.7.2/lib/python3.7/contextlib.py in __exit__(self, type, value, traceback)
    128                 value = type()
    129             try:
--> 130                 self.gen.throw(type, value, traceback)
    131             except StopIteration as exc:
    132                 # Suppress StopIteration *unless* it's the same exception that

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    623         from numba import config
    624         tb = sys.exc_info()[2] if config.FULL_TRACEBACKS else None
--> 625         six.reraise(type(newerr), newerr, tb)
    626 
    627 

~/.local/share/virtualenvs/poker-ai-zNaDkIqV/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    657         if value.__traceback__ is not tb:
    658             raise value.with_traceback(tb)
--> 659         raise value
    660 
    661 else:

LoweringError: Failed in nopython mode pipeline (step: nopython mode backend)
Cannot cast array(float64, 1d, C) to bool: %".73" = load {i8*, i8*, i64, i64, double*, [1 x i64], [1 x i64]}, {i8*, i8*, i64, i64, double*, [1 x i64], [1 x i64]}* %"bar"

File "<ipython-input-3-127d50b78b35>", line 4:
def truth_value():
    <source elided>
    bar = np.zeros(4)
    if bar:
    ^

[1] During: lowering "branch bar, 14, 26" at <ipython-input-3-127d50b78b35> (4)
-------------------------------------------------------------------------------
This should not have happened, a problem has occurred in Numba's internals.

Please report the error message and traceback, along with a minimal reproducer
at: https://github.com/numba/numba/issues/new

If more help is needed please feel free to speak to the Numba core developers
directly at: https://gitter.im/numba/numba

Thanks in advance for your help in improving Numba!
````

## Reporting a bug

- [x] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

**Comment by sklam** [comment]
The reason for the error is that there is no implementation for turning an array to boolean.  The operation is not allowed in NumPy.  When the code is run without `@njit`, you will get: 

```
<ipython-input-3-8758756a6ca2> in truth_value()
      5 def truth_value():
      6     bar = np.zeros(4)
----> 7     if bar:
      8         bar[1] = 1
      9         return bar

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

Even though Numba has correctly raised an exception for the unsupported error, it should generate better error message for this.  The typing phase should have caught this error.

**Comment by cstich** [comment]
That makes sense. I just did what the traceback told me to and reported this :-)

**Comment by sklam** [comment]
Note from meeting: should add a `truth()` to explicit convert values into boolean at the branch before hitting the LLVM.

**Comment by stuartarchibald** [comment]
#3901 fixed and as of mainline for 0.55, Numba produces:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
which matches NumPy. Closing as fixed. Thanks!
